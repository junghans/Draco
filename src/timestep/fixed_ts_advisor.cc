//----------------------------------*-C++-*----------------------------------//
/*!
 * \file   fixed_ts_manager.cc
 * \author John McGhee
 * \date   Mon Apr  6 17:22:53 1998
 * \brief  Defines the fixed time-step advisor.
 * \note   Copyright Â© 1998-2010 Los Alamos National Security, LLC.
 */
//---------------------------------------------------------------------------//
// $Id$
//---------------------------------------------------------------------------//

#include "fixed_ts_advisor.hh"
#include "ds++/Assert.hh"
#include "c4/global.hh"
#include <iostream>

using std::cout;
using std::endl;

namespace rtt_timestep
{

fixed_ts_advisor::fixed_ts_advisor(const std::string &name_,
				   const usage_flag usage_,
				   const double fixed_value_,
				   const bool active_)

    : ts_advisor(name_, usage_, active_),
      fixed_value(fixed_value_)
{
    Ensure(invariant_satisfied());
}


fixed_ts_advisor::~fixed_ts_advisor()
{
// empty
}

double fixed_ts_advisor::get_dt_rec(const ts_manager &/*tsm*/) const
{
    Require(invariant_satisfied());
    return fixed_value;
}

void fixed_ts_advisor::print_state() const
{
    if (C4::node() != 0)
	return;
    
    std::string status = is_active() ? "true " : "false";
    cout << endl;
    cout << "  ** Time-Step Advisor State Listing **" << endl;
    cout << "  Name - " << get_name() << endl;
    cout << "  Type           : " << "Fixed Advisor" << endl;
    cout << "  Active         : " << status << endl;
    cout << "  Usage          : " << usage_flag_name(get_usage()) << endl;
    cout << "  Fixed Value    : " << fixed_value << endl;
    cout << endl;
}

bool fixed_ts_advisor::invariant_satisfied() const
{
    return ts_advisor::invariant_satisfied() &&
        0. < fixed_value;
}

} //end of rtt_timestep namespace

//---------------------------------------------------------------------------//
//                              end of fixed_ts_advisor.cc
//---------------------------------------------------------------------------//
