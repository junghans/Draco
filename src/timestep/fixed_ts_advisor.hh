//----------------------------------*-C++-*----------------------------------//
// fixed_ts_advisor.hh
// John McGhee
// Thu Apr  2 14:06:18 1998
//---------------------------------------------------------------------------//
// @> 
//---------------------------------------------------------------------------//

#ifndef __timestep_fixed_ts_advisor_hh__
#define __timestep_fixed_ts_advisor_hh__

//===========================================================================//
// class fixed_ts_advisor - This class provides a means to introduce
// a user defined fixed value into the time-step calculation.
// This is useful to set min and max timesteps, or to force a
// timestep, etc.
// 
//===========================================================================//

#include "timestep/ts_advisor.hh"

class fixed_ts_advisor : public ts_advisor {

  public:

    fixed_ts_advisor( 
	const std::string &name_  = std::string("Unlabeled"),
	const usage_flag usage_ = max, 
	const double const_value_ = large(),
	const bool active_ = true);
    
    ~fixed_ts_advisor();

// Update the recommended time-step
    
    void update_tstep(const int cycle_);

// Set the fixed value

    void set_fixed_value(const double value_ = large())
    { 
	fixed_value = value_;
    }

// Print state

    void print_state();

  private:

    bool invariant_satisfied();

    double fixed_value;     //value used to oompute fixed advisor 

};

#endif                          // __timestep_fixed_ts_advisor_hh__

//---------------------------------------------------------------------------//
//                              end of timestep/fixed_ts_advisor.hh
//---------------------------------------------------------------------------//
