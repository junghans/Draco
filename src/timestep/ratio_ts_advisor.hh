//----------------------------------*-C++-*----------------------------------//
// ratio_ts_advisor.hh
// John McGhee
// Thu Apr  2 14:06:18 1998
//---------------------------------------------------------------------------//
// @> Defines the ratio time-step advisor.
//---------------------------------------------------------------------------//

#ifndef __timestep_ratio_ts_advisor_hh__
#define __timestep_ratio_ts_advisor_hh__

//===========================================================================//
// class ratio_ts_advisor - Calculates a new timestep as a ratio of 
//                          the current time-step
//
// This class provides a means to calculate a
// new timestep as a ratio of the current time-step. This is useful to
// limit the rate of change in the time-step from one time cycle to the 
// next.
// 
//===========================================================================//

#include "timestep/ts_advisor.hh"

class ratio_ts_advisor : public ts_advisor {


// DATA

 // Value used to compute ratio advisor

  private:
    double ratio_value; 



// CREATORS
  public:

    ratio_ts_advisor(const std::string &name_  = std::string("Unlabeled"),
		     const usage_flag usage_ = max,
		     const double ratio_value_ = 1.20,
		     const bool active_ = true);
    
    ~ratio_ts_advisor();



// MANIPULATORS

// Update the recommended time-step 
    
    void update_tstep(const double current_dt,
		      const int cycle_);

// Set the ratio value

    void set_ratio(const double value_ = 1.2)
    { 
	ratio_value = value_;
    }



// ACCESSORS

// Print state

    void print_state() const;

  private:

    bool invariant_satisfied() const;


};

#endif                          // __timestep_ratio_ts_advisor_hh__

//---------------------------------------------------------------------------//
//                              end of timestep/ratio_ts_advisor.hh
//---------------------------------------------------------------------------//
