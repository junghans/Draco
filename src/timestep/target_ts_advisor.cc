//----------------------------------*-C++-*----------------------------------//
// target_ts_advisor.cc
// John McGhee
// Thu Apr  2 14:06:18 1998
//---------------------------------------------------------------------------//
// @> Defines the target time-step advisor.
//---------------------------------------------------------------------------//

#include "timestep/target_ts_advisor.hh"

#include "ds++/Assert.hh"

#include <iostream>

using std::cout;
using std::endl;

target_ts_advisor::target_ts_advisor(
    const std::string   &name_,
    const usage_flag usage_,
    const double target_value_,
    const bool active_)

    : ts_advisor(name_, usage_, active_),
      target_value(target_value_)

{
    Ensure(invariant_satisfied());
}

target_ts_advisor::~target_ts_advisor()
{
// empty
}

void target_ts_advisor::update_tstep( const double end_of_cycle_time,
				      const int cycle_)
{
    Require(invariant_satisfied());
    
    dt_rec = target_value - end_of_cycle_time;
    if (dt_rec <= small ())
    {
	dt_rec = large();
    }
    
    cycle_at_last_update = cycle_;
    Ensure(invariant_satisfied());
}

void target_ts_advisor::print_state() const
{
    std::string status = active ? "true " : "false";
    cout << endl;
    cout << "  ** Time-Step Advisor State Listing **" << endl;
    cout << "  Name - " << name << endl;
    cout << "  Type           : " << "Target Advisor" << endl;
    cout << "  Active         : " << status << endl;
    cout << "  Usage          : " << usage_flag_name(usage) << endl;
    cout << "  Last Update    : " << "cycle " << cycle_at_last_update << endl;
    cout << "  Target Value   : " << target_value << endl;
    cout << "  dt_rec         : " << dt_rec << endl;
    cout << endl;
}

bool target_ts_advisor::invariant_satisfied() const
{
    bool ldum =
	name.length() != 0 &&
	0      <= usage &&
	usage  <  last_usage &&
	0. < dt_rec;

    return ldum;
}

//---------------------------------------------------------------------------//
//                              end of target_ts_advisor.cc
//---------------------------------------------------------------------------//
