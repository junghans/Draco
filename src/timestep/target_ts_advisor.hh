//----------------------------------*-C++-*----------------------------------//
// target_ts_advisor.hh
// John McGhee
// Thu Apr  2 14:06:18 1998
//---------------------------------------------------------------------------//
// @> Defines the target time-step advisor.
//---------------------------------------------------------------------------//

#ifndef __timestep_target_ts_advisor_hh__
#define __timestep_target_ts_advisor_hh__

//===========================================================================//
// class target_ts_advisor - Calculates a new time-step required to hit
//                           some "target" problem time.
//
// This class provides a means to calculate 
// a new time-step required to hit some "target" problem time. This
// is useful to assure that graphics dumps/IO/etc. occur precisely at
// a predetermined problem time.
// 
//===========================================================================//

#include "timestep/ts_advisor.hh"

class target_ts_advisor : public ts_advisor {

// DATA

// Target time (time)

  private:
    double target_value; 


// CREATORS

  public:
    
    target_ts_advisor( 
	const std::string &name_  = std::string("Unlabeled"),
	const usage_flag usage_ = max,
	const double target_value_ = -large(),
	const bool active_ = true );
    
    ~target_ts_advisor();



// MANIPULATORS

// Update the recommended time-step 
    
    void update_tstep( const double end_of_cycle_time,
		       const int cycle_ );

// Set the target value

    void set_target(const double value_ = -large() )
    { 
	target_value = value_;
    }



// ACCESSORS

// Print state

    void print_state() const;

  private:

    bool invariant_satisfied() const;


};

#endif                          // __target_timestep_ts_advisor_hh__

//---------------------------------------------------------------------------//
//                              end of timestep/target_ts_advisor.hh
//---------------------------------------------------------------------------//
