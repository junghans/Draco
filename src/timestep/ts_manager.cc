//----------------------------------*-C++-*----------------------------------//
// ts_manager.cc
// John McGhee
// Mon Apr  6 17:22:53 1998
//---------------------------------------------------------------------------//
// @> 
//---------------------------------------------------------------------------//

#include "timestep/ts_manager.hh"

#include "ds++/Assert.hh"

#include <functional>

#include <stdexcept>

#include <iostream>

using std::list;

typedef ts_advisor TSA;

using std::endl;
using std::cerr;
using std::cout;
using std::ios;

ts_manager::ts_manager()
    :time(0.), dt_new(TSA::small()), dt(TSA::small()), cycle(9999),  
     controlling_advisor("Not Set")
{
    Ensure(invariant_satisfied());
}

ts_manager::~ts_manager()
{
// empty
}

bool ts_manager::invariant_satisfied()
{
    bool ldum =
	0.0 < dt_new &&
	0.0 < dt &&
	controlling_advisor.length() != 0;
    return ldum;
}

double ts_manager::compute_new_timestep(double dt_, int cycle_,
					double time_)
{

    cycle = cycle_;
    dt    = dt_;
    time  = time_;

// Check to be sure that there is at least one usable advisor

    bool found = false;
    for (list< SP<ts_advisor> >::iterator py = advisors.begin(); 
	 py != advisors.end(); py++) 
    {
	if ((**py).advisor_usable(cycle))
	{
	    found = true;
	    break;
	}
    }
    
    if (!found)
    {
	cerr << "  ** Time-Step Manager Warning **" << endl;
	cerr << "  No usable time-step advisors found," << endl;
	cerr << "  defaulting to current time-step" << endl;
	dt_new = dt_;
	controlling_advisor = "Current Time-Step";
	return dt_new;
    }

// Check for one and only one mandatory advisor

    int i = 0;
    for (list< SP<ts_advisor> >::iterator py = advisors.begin(); 
	 py != advisors.end(); py++) 
    {
	if ((**py).advisor_usable(cycle) && (**py).get_usage() == TSA::req)
	{
	    i++;
	    dt_new = (**py).get_dt_rec();
	    controlling_advisor = (**py).get_name();
	}
    } 
    if (i == 1)
    {
	return dt_new;
    }
    else if (i != 0)
    {
	cerr << "  ** Time-Step Manager Warning **" << endl;
	cerr << "  Cycle Number: " << cycle << endl;
	cerr << "  More than one mandatory advisor found," << endl;
	cerr << "  defaulting to last found" << endl;
    }

// Loop over the advisors finding the one that controls

    list< SP<ts_advisor> >::iterator py1=advisors.end();
    list< SP<ts_advisor> >::iterator py2=advisors.end();
    double x1 = TSA::small();
    double x2 = TSA::large();
    for (list< SP<ts_advisor> >::iterator py = advisors.begin(); 
	 py != advisors.end(); py++) 
    {   
	if ((**py).advisor_usable(cycle))
	{
            if ((**py).get_usage() == TSA::min)
	    {
		if ((**py).get_dt_rec() > x1)
		{
		    x1 = (**py).get_dt_rec();
		    py1 = py;
		}
	    }
	    else if ((**py).get_usage() == TSA::max)
	    {
		if ((**py).get_dt_rec() <  x2)
		{
		    x2 = (**py).get_dt_rec();
		    py2 = py;
		}
	    }
	}
    }

    if (py1 == advisors.end() && py2 == advisors.end() )
    {
	cerr << "  ** Time-Step Manager Warning **" << endl;
	cerr << "  Cycle Number: " << cycle << endl;
	cerr << "  No usable time-step advisors found," << endl;
	cerr << "  defaulting to current time-step" << endl;
	dt_new = dt_;
	controlling_advisor = "Current Time-Step";
	return dt_new;
    }
    else if (py1 == advisors.end())
    {
	dt_new = x2;
	controlling_advisor = (**py2).get_name();
    }
    else if (py2 == advisors.end())
    {
	dt_new = x1;
	controlling_advisor = (**py1).get_name();
    }
    else
    {
	if (x1 > x2)
	{
	    dt_new = x1;
	    controlling_advisor = (**py1).get_name();
	    cerr << "  ** Time-Step Manager Warning **" << endl;
	    cerr << "  Cycle Number: " << cycle << endl;
	    cerr << "  No window between min and max advisors," << endl;
	    cerr << "  defaulting to min recommended dt" << endl;
	}
	else
	{
	    dt_new = x2;
	    controlling_advisor = (**py2).get_name();
	}
    }
    
    return dt_new;
}

void ts_manager::add_advisor(const SP<ts_advisor> &new_advisor)
{
    bool not_in_use = true;
    for (list< SP <ts_advisor> >::iterator py = advisors.begin(); 
	 py != advisors.end(); py++) 
    {
	if ((**py).get_name() == (*new_advisor).get_name())
	{
	    not_in_use = false;
	    break;
	}
    }
    if (not_in_use)
    {
	advisors.push_front(new_advisor);
    }
    else
    {
	throw std::runtime_error("Name for requested advisor already in use");
    }
}

void ts_manager::remove_advisor( const SP<ts_advisor> &advisor_to_remove)
{
    for (list< SP<ts_advisor> >::iterator py = advisors.begin(); 
	 py != advisors.end(); py++) 
    {
	if ((**py).get_name() == (*advisor_to_remove).get_name())
	{
	    advisors.erase(py);
	    return;
	}
    }
    throw std::runtime_error("Unable to find requested advisor");
}


struct sptsa_less_than : public std::binary_function< SP<ts_advisor>,
			 SP<ts_advisor>, bool > 
{
    bool operator () (const SP<ts_advisor> &sp_lhs, 
		      const SP<ts_advisor> &sp_rhs) const
    {
	return (*sp_lhs < *sp_rhs);
    }
};

void ts_manager::print_summary() const
{
    cout.setf(ios::scientific, ios::floatfield);
    cout.precision(4);
    list < SP<ts_advisor> > temp = advisors;
    temp.sort(sptsa_less_than());
    cout << endl;
    cout << "  *** Time-Step Manager Summary ***" << endl;
    cout << "  Cycle Number         : " << cycle << endl;
    cout << "  Problem Time         : " << time << endl;
    cout << "  Current Time-Step    : " << dt << endl;
    cout << "  Recommended Time-Step: " << dt_new << endl;
    cout << "  Controlling Advisor  : " << controlling_advisor << endl;
    cout << endl;
    cout << "  *** Time-Step Advisor Data Table *** " << endl;
    cout << "  Recommendation   In-Use  Name " << endl;
    for (list< SP<ts_advisor> >::const_iterator py = temp.begin(); 
	 py != temp.end(); py++)
    {
	(**py).print(cycle, (**py).get_name() == controlling_advisor);
    }
    cout << endl;
    cout.setf(0,ios::floatfield);
}

void ts_manager::print_advisors() const
{
    cout << endl;
    cout << "*** Time-Step Manager: Advisor Listing ***" << endl;
    for (list< SP<ts_advisor> >::const_iterator 
	     py = advisors.begin(); py != advisors.end(); py++)
    {
	cout << (**py).get_name() << endl;
    }
    cout << endl; 
}


//---------------------------------------------------------------------------//
//                              end of ts_manager.cc
//---------------------------------------------------------------------------//
