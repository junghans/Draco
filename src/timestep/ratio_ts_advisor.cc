//----------------------------------*-C++-*----------------------------------//
/*!
 * \file   ratio_ts_advisor.cc
 * \author John McGhee
 * \date   Thu Apr  2 14:06:18 1998
 * \brief  Defines the ratio time-step advisor.
 * \note   Copyright (C) 1998-2010 Los Alamos National Security, LLC.
 *         All rights reserved.
 * \version $Id$
 */
//---------------------------------------------------------------------------//

#include "ratio_ts_advisor.hh"
#include "ts_manager.hh"
#include "ds++/Assert.hh"
#include "ds++/config.h"
#include "c4/global.hh"
#include <iostream>

using std::cout;
using std::endl;

namespace rtt_timestep
{

ratio_ts_advisor::ratio_ts_advisor(const std::string &name_,
				   const usage_flag usage_,
				   const double ratio_value_,
				   const bool active_) 

    : ts_advisor(name_, usage_, active_), ratio_value(ratio_value_)

{
    Ensure(invariant_satisfied());
}


ratio_ts_advisor::~ratio_ts_advisor()
{
// empty
}

double ratio_ts_advisor::get_dt_rec(const ts_manager &tsm) const
{
    Require(invariant_satisfied());
    Require(tsm.get_dt() > 0.);
    return ratio_value*tsm.get_dt();
}

void ratio_ts_advisor::print_state() const
{
    if (C4::node() != 0)
	return;
    
    std::string status = is_active() ? "true " : "false";
    cout << endl;
    cout << "  ** Time-Step Advisor State Listing **" << endl;
    cout << "  Name - " << get_name() << endl;
    cout << "  Type           : " << "Ratio Advisor" << endl;
    cout << "  Active         : " << status << endl;
    cout << "  Usage          : " << usage_flag_name(get_usage()) << endl;
    cout << "  Ratio Value    : " << ratio_value << endl;
    cout << endl;
}

bool ratio_ts_advisor::invariant_satisfied() const
{
    return ts_advisor::invariant_satisfied() &&
        0. < ratio_value;
}

} //end of rtt_timestep namespace

//---------------------------------------------------------------------------//
//                              end of ratio_ts_advisor.cc
//---------------------------------------------------------------------------//
