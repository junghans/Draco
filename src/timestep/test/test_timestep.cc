//----------------------------------*-C++-*----------------------------------//
// test_timestep.cc
// John McGhee
// Fri May  1 09:44:46 1998
//---------------------------------------------------------------------------//
// @> Performs test of time-step manager facility.
//---------------------------------------------------------------------------//

#include "timestep/test/test_timestep.hh"

#include "timestep/ts_manager.hh"
#include "timestep/fixed_ts_advisor.hh"
#include "timestep/ratio_ts_advisor.hh"
#include "timestep/target_ts_advisor.hh"
#include "timestep/test/dummy_package.hh"
#include "timestep/test/test_utils.hh"

#include <iostream>
#include <cmath>

using std::cout;
using std::endl;

test_timestep::test_timestep()
{
//empty
}

test_timestep::~test_timestep()
{
//empty
}

void test_timestep::execute_test()
{

    using dsxx::SP;

    double graphics_time = 10.;
    double dt_min = 0.000001;
    double dt_max = 100000.;
    bool override_flag = false;
    double override_dt = 1.;
    int icycle_first = 1;
    int icycle_last = 3;
    double dt = 1.;
    double time = 0.;
    ts_manager mngr;
    dummy_package xxx(mngr);

// Set up a informational advisor to
// contain the current time-step for reference.
// Activating this controller can also be used to
// freeze the time-step at the current value.

    SP<fixed_ts_advisor> sp_dt 
	(new fixed_ts_advisor("Current Time-Step",
			      ts_advisor::req, dt, false));
    mngr.add_advisor(sp_dt);

// Set up a required time-step to be activated
// at the user's discretion

    SP<fixed_ts_advisor> sp_ovr 
	(new fixed_ts_advisor("User Override",
			      ts_advisor::req, 
			      override_dt, false));
    mngr.add_advisor(sp_ovr);

// Set up a min timestep

    SP<fixed_ts_advisor> sp_min 
	(new fixed_ts_advisor("Minimum",
			      ts_advisor::min, 
			      ts_advisor::small()));
    mngr.add_advisor(sp_min);
    sp_min -> set_fixed_value(dt_min);

// Set up a lower limit on the timestep rate of change

    SP<ratio_ts_advisor> sp_llr 
	(new ratio_ts_advisor( "Rate of Change Lower Limit",
			       ts_advisor::min, 0.8 ) );
    mngr.add_advisor(sp_llr);

// Set up an upper limit on the time-step rate of change

    SP<ratio_ts_advisor> sp_ulr 
	(new ratio_ts_advisor("Rate of Change Upper Limit"));
    mngr.add_advisor(sp_ulr);

// Set up an advisor to watch for an upper limit on the time-step.

    SP<fixed_ts_advisor> sp_max (new fixed_ts_advisor("Maximum"));
    mngr.add_advisor(sp_max);
    sp_max -> set_fixed_value(dt_max);

// Set up a target time advisor

    SP<target_ts_advisor> sp_gd 
	(new target_ts_advisor("Graphics Dump",
			       ts_advisor::max, graphics_time));
    mngr.add_advisor(sp_gd);

// Now that all the advisors have been set up, perform time cycles

    for (int i=icycle_first; i != icycle_last+1; i++)
    {

	time = time + dt; //end of cycle time
	mngr.set_cycle_data(dt,i,time);

    // Make any user directed changes to controllers

	sp_dt -> set_fixed_value(dt);
	if (override_flag)
	{
	    sp_ovr -> activate();
	    sp_ovr -> set_fixed_value(override_dt);
	}
	else
	{
	    sp_ovr -> deactivate();
	}

    // Pass in the advisors owned by package_XXX for
    // that package to update

   	xxx.advance_state();

    //Compute a new time-step and print results to screen

	dt = mngr.compute_new_timestep();
	mngr.print_summary();
	    
    }

// Dump a list of the advisors to the screen

    mngr.print_advisors();

// Dump the advisor states for visual examination.

    mngr.print_adv_states();

// Confirm that at least some of the output is correct.

    int nd = 5;
    bool passed = mngr.get_cycle() == icycle_last
	&& compare_reals(3.345679e+00, mngr.get_time(),nd)
	&& compare_reals(1.234568e+00, mngr.get_dt(),nd)
	&& compare_reals(1.371742e+00, mngr.get_dt_new(),nd)
	&& mngr.get_controlling_advisor() == "Electron Temperature"
	&& compare_reals(1.000000e-06,sp_min->get_dt_rec(mngr),nd)
	&& compare_reals(9.876543e-01,sp_llr->get_dt_rec(mngr),nd)
	&& compare_reals(1.000000e+00,sp_ovr->get_dt_rec(mngr),nd)
	&& compare_reals(1.234568e+00, sp_dt->get_dt_rec(mngr),nd)
	&& compare_reals(1.481481e+00,sp_ulr->get_dt_rec(mngr),nd)
	&& compare_reals(6.654321e+00, sp_gd->get_dt_rec(mngr),nd)
	&& compare_reals(1.000000e+05,sp_max->get_dt_rec(mngr),nd)
	&& xxx.tests_passed();
	
// Print the status of the test.

    cout << endl;
    cout <<     "*********************************************" << endl;
    if (passed) 
    {
	cout << "**** Time-step Manager Self Test: PASSED ****" << endl;
    }
    else
    {
	cout << "**** Time-step Manager Self Test: FAILED ****" << endl;
    }
    cout <<     "*********************************************" << endl;
    cout << endl;
}


//---------------------------------------------------------------------------//
//                              end of test_timestep.cc
//---------------------------------------------------------------------------//
