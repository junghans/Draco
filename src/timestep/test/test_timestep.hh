//----------------------------------*-C++-*----------------------------------//
// test_timestep.hh
// John McGhee
// Fri May  1 09:44:46 1998
//---------------------------------------------------------------------------//
// @> Tests the time-step manager and advisors.
//---------------------------------------------------------------------------//

#ifndef __timestep_test_test_timestep_hh__
#define __timestep_test_test_timestep_hh__

//===========================================================================//
// class test_timestep -  Tests the time-step manager and advisors.

// 
//===========================================================================//

class test_timestep {

  public:
    test_timestep();
    ~test_timestep();

    void execute_test();
};

#endif                          // __timestep_test_test_timestep_hh__

//---------------------------------------------------------------------------//
//                              end of test_timestep.hh
//---------------------------------------------------------------------------//
