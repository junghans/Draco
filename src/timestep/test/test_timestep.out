
  *** Time-Step Manager Summary ***
  Cycle Number         : 1
  Problem Time         : 1.0000e+00
  Current Time-Step    : 1.0000e+00
  Recommended Time-Step: 1.1111e+00
  Controlling Advisor  : Electron Temperature

  *** Time-Step Advisor Data Table *** 
  Recommendation   In-Use  Name 
      1.0000e-06   true    Minimum
      8.0000e-01   true    Rate of Change Lower Limit
      1.0000e+00   false   User Override
      1.0000e+00   false   Current Time-Step
  ==> 1.1111e+00   true    Electron Temperature
      1.2000e+00   true    Rate of Change Upper Limit
      1.2125e+00   true    Ion Temperature
      2.2000e+00   true    Radiation Intensity
      9.0000e+00   true    Graphics Dump
      1.0000e+05   true    Maximum


  *** Time-Step Manager Summary ***
  Cycle Number         : 2
  Problem Time         : 2.1111e+00
  Current Time-Step    : 1.1111e+00
  Recommended Time-Step: 1.2346e+00
  Controlling Advisor  : Electron Temperature

  *** Time-Step Advisor Data Table *** 
  Recommendation   In-Use  Name 
      1.0000e-06   true    Minimum
      8.8889e-01   true    Rate of Change Lower Limit
      1.0000e+00   false   User Override
      1.1111e+00   false   Current Time-Step
  ==> 1.2346e+00   true    Electron Temperature
      1.3333e+00   true    Rate of Change Upper Limit
      1.3472e+00   true    Ion Temperature
      2.4444e+00   true    Radiation Intensity
      7.8889e+00   true    Graphics Dump
      1.0000e+05   true    Maximum


  *** Time-Step Manager Summary ***
  Cycle Number         : 3
  Problem Time         : 3.3457e+00
  Current Time-Step    : 1.2346e+00
  Recommended Time-Step: 1.3717e+00
  Controlling Advisor  : Electron Temperature

  *** Time-Step Advisor Data Table *** 
  Recommendation   In-Use  Name 
      1.0000e-06   true    Minimum
      9.8765e-01   true    Rate of Change Lower Limit
      1.0000e+00   false   User Override
      1.2346e+00   false   Current Time-Step
  ==> 1.3717e+00   true    Electron Temperature
      1.4815e+00   true    Rate of Change Upper Limit
      1.4969e+00   true    Ion Temperature
      2.7160e+00   true    Radiation Intensity
      6.6543e+00   true    Graphics Dump
      1.0000e+05   true    Maximum


*** Time-Step Manager: Advisor Listing ***
Graphics Dump
Maximum
Rate of Change Upper Limit
Rate of Change Lower Limit
Minimum
User Override
Current Time-Step
Radiation Intensity
Ion Temperature
Electron Temperature


*** Time-Step Manager: Advisor State Listing ***

  ** Time-Step Advisor State Listing **
  Name - Graphics Dump
  Type           : Target Advisor
  Active         : true 
  Usage          : maximum
  Target Value   : 10


  ** Time-Step Advisor State Listing **
  Name - Maximum
  Type           : Fixed Advisor
  Active         : true 
  Usage          : maximum
  Fixed Value    : 1e+05


  ** Time-Step Advisor State Listing **
  Name - Rate of Change Upper Limit
  Type           : Ratio Advisor
  Active         : true 
  Usage          : maximum
  Ratio Value    : 1.2


  ** Time-Step Advisor State Listing **
  Name - Rate of Change Lower Limit
  Type           : Ratio Advisor
  Active         : true 
  Usage          : minimum
  Ratio Value    : 0.8


  ** Time-Step Advisor State Listing **
  Name - Minimum
  Type           : Fixed Advisor
  Active         : true 
  Usage          : minimum
  Fixed Value    : 1e-06


  ** Time-Step Advisor State Listing **
  Name - User Override
  Type           : Fixed Advisor
  Active         : false
  Usage          : required
  Fixed Value    : 1


  ** Time-Step Advisor State Listing **
  Name - Current Time-Step
  Type           : Fixed Advisor
  Active         : false
  Usage          : required
  Fixed Value    : 1.235


  ** Time-Step Advisor State Listing **
  Name - Radiation Intensity
  Type           : Field Advisor
  Active         : true 
  Usage          : maximum
  Last Update    : cycle 3
  Update Method  : field value and relative change
  Fract Change   : 0.1
  Floor Value    : 2.225e-306
  dt_rec         : 2.716


  ** Time-Step Advisor State Listing **
  Name - Ion Temperature
  Type           : Field Advisor
  Active         : true 
  Usage          : maximum
  Last Update    : cycle 3
  Update Method  : weighted by relative change
  Fract Change   : 0.1
  Floor Value    : 2.225e-306
  dt_rec         : 1.497


  ** Time-Step Advisor State Listing **
  Name - Electron Temperature
  Type           : Field Advisor
  Active         : true 
  Usage          : maximum
  Last Update    : cycle 3
  Update Method  : arithmetic mean
  Fract Change   : 0.1
  Floor Value    : 0.01199
  dt_rec         : 1.372



*********************************************
**** Time-step Manager Self Test: PASSED ****
*********************************************


